packages/db-migrations/databases/fxa/patches/patch-134-135.sql

```diff
@@ -0,0 +1,112 @@
+SET NAMES utf8mb4 COLLATE utf8mb4_bin;
+
+CALL assertPatchLevel('134');
+
+
+-- This migration adds a routine for removing sessions on accounts that have
+-- over maxSessions. Essentially, per account tokens above maxSessions
+-- will be purged. Older tokens will be removed first. Associated unverified
+-- tokens and devices will also be removed.
+--   maxSessions - Max number of sessions allowed per account
+--   accountLimit - Max number of accouts to process at a time
+--   deletionLimit - Max number of session tokens to delete at a time
+--   accountsOverLimit - Number of accounts having more than the allowed session token limit
+--   totalDeletions - Total number of rows deleted, which includes session token rows, unverified session token rows and device rows.
+CREATE PROCEDURE `limitSessions_2` (
+  IN `maxSessions` INT UNSIGNED,
+  IN `accountLimit` INT UNSIGNED,
+  IN `deletionLimit` INT UNSIGNED,
+  OUT `accountsOverLimit` INT UNSIGNED,
+  OUT `totalDeletions` INT UNSIGNED
+)
+BEGIN
+  DECLARE EXIT HANDLER FOR SQLEXCEPTION
+  BEGIN
+    ROLLBACK;
+    RESIGNAL;
+  END;
+
+  SET accountsOverLimit = 0;
+  SET totalDeletions = 0;
+
+  -- We will share a lock between this procedure and the prune_* procedure. The code should
+  -- invoke these sequentially, but this is an extra guard.
+  SELECT @lockAcquired := GET_LOCK('fxa-auth-server.prune-lock', 3);
+  IF @lockAcquired THEN
+
+    -- First create a temp table containing just the accounts having too many sessions.
+    -- In practice, this should be a small percentage of all accounts, assuming that
+    -- maxSessions is adequately large. This let's subsequent queries efficiently target
+    -- a subset of data needs to be deleted. Furhtermore this query is limitted to maxSessions
+    -- accounts to constrain the scope of the delete.
+    DROP TEMPORARY TABLE IF EXISTS tokenCounts;
+    CREATE TEMPORARY TABLE tokenCounts
+      SELECT uid
+      FROM (
+        SELECT uid, COUNT(tokenId) AS tokenCount
+        FROM sessionTokens
+        GROUP BY uid
+      ) AS counts
+      WHERE tokenCount > maxSessions
+      ORDER BY counts.tokenCount DESC
```

Comment:
@chenba I just added this to make sure we are really addressing the accounts with the highest session counts first. This `ORDER BY` probably adds some expense though...